p2panda-rs/src/secret_group/group.rs

```diff
@@ -0,0 +1,552 @@
+    /// MLS group TreeKEM algorithm.
+    ///
+    /// This method gives forward-secrecy and post-compromise security. Use this encryption method
+    /// for highly secure group settings and applications where decrypted data can safely be stored
+    /// on the clients device since secrets are meant to be thrown away and not be reused, past
+    /// data can not be recovered.
+    pub fn encrypt(
+        &mut self,
+        provider: &impl OpenMlsCryptoProvider,
+        data: &[u8],
+    ) -> Result<SecretGroupMessage, SecretGroupError> {
+        let mls_ciphertext = self.mls_group.encrypt(provider, data)?;
+        Ok(SecretGroupMessage::MlsApplicationMessage(mls_ciphertext))
+    }
+
+    /// Encrypt user data using the group's current symmetrical long term secret.
+    ///
+    /// This method gives only post-compromise security and has in general lower security
+    /// guarantees but gives more flexibility. Use this encryption method if you want every old or
+    /// new group member to decrypt past data even when they've joined the group later.
+    pub fn encrypt_with_long_term_secret(
+        &mut self,
+        provider: &impl OpenMlsCryptoProvider,
+        data: &[u8],
+    ) -> Result<SecretGroupMessage, SecretGroupError> {
+        // Generate unique nonce for AEAD encryption
+        let nonce = self.generate_nonce(provider)?;
+
+        // Unwrap here since at this stage we already have at least one LTS epoch
+        let epoch = self.long_term_epoch().unwrap();
+        let secret = self.long_term_secret(epoch).unwrap();
+
+        // Encrypt user data with last long term secret
+        let ciphertext = secret.encrypt(provider, &nonce, data)?;
+
+        Ok(SecretGroupMessage::LongTermSecretMessage(ciphertext))
+    }
+
+    /// Decrypts user data.
+    ///
+    /// This method automatically detects if the ciphertext was encrypted with a Sender Ratchet
+    /// Secret or a Long Term Secret and returns an `SecretGroupError` if the required key material
+    /// for decryption is missing.
+    pub fn decrypt(
+        &mut self,
+        provider: &impl OpenMlsCryptoProvider,
+        message: &SecretGroupMessage,
+    ) -> Result<Vec<u8>, SecretGroupError> {
+        match message {
+            SecretGroupMessage::MlsApplicationMessage(ciphertext) => {
+                Ok(self.mls_group.decrypt(provider, ciphertext.clone())?)
+            }
+            SecretGroupMessage::LongTermSecretMessage(ciphertext) => {
+                let secret = self
+                    .long_term_secret(ciphertext.long_term_epoch())
+                    .ok_or_else(|| SecretGroupError::LTSSecretMissing)?;
+                Ok(secret.decrypt(provider, ciphertext)?)
+            }
+        }
+    }
+
+    // Status
+    // ======
+
+    /// Returns true if this group is still active or false if the member got removed from the
+    /// group.
+    pub fn is_active(&self) -> bool {
```

Comment:
An MLS group member is a public key, it will never understand what a `KeyGroup` is :smile_cat: If we want a `KeyGroup` in an MLS group then every key pair needs to create an `SecretGroupMember` and `KeyPackage` and needs to be explicitly added.

So `is_active` will return if you are still a member of the group in the scope of your one public key your client is currently using. If we are interested in addressing this for whole key groups this would need to be implemented on top of this method.

Reply:
Completely independent from key groups. I was just trying to understand in what context `is_active` is operating. 

I originally wrote this comment before reading through all the tests. I now understand that a `SecretGroup` instance is always bound to a specific `KeyPair`. That was not immediately obvious to me and answers this question.

I think it would be very helpful to add this information to the `SecretGroup` docstring.

Reply:
Cool! I've added a note there, maybe you can have a look if its understandable :sweat_smile: